### PR TITLE
Plugin build changes

### PR DESCRIPTION
**Pull request by zachmullen**
This is the last major build system change prior to 2.0 release. This change makes it so that we no longer build all plugins in the plugins directory by default when running `npm run build` or `grunt`. To build plugins, one can now either run `npm run build -- --plugins=x,y,z` or simply build all available plugins as before via `npm run build -- --all-plugins`.

These commands will also be exposed behind new `girder-install web` commands, which by default will now build all *enabled* plugins rather than all available ones.

Finally, the ability to rebuild plugins as well as the core web client code will be exposed via the admin console to make this as painless as possible for site maintainers.

### CODE REVIEW

**Comment by zachmullen** [comment]
@girder/developers this is now ready for review.

**Comment by zachmullen** [comment]
Does anybody know how it's possible that this is consistently failing on push and consistently passing on PR even after I rebase onto the latest 2.0-integration? The failure message on the head of the branch is:

```
BABEL_ENV=cover girder-install web --all-plugins --dev
girder@1.6.0 /home/travis/build/girder/girder
├─┬ UNMET PEER DEPENDENCY webpack@2.1.0-beta.21
│ ├─┬ node-libs-browser@1.0.0
│ │ └─┬ crypto-browserify@3.11.0
│ │   ├─┬ browserify-sign@4.0.0
│ │   │ └─┬ parse-asn1@5.0.0
│ │   │   └── asn1.js@4.8.1 
│ │   └── pbkdf2@3.0.8 
│ └─┬ watchpack@1.1.0
│   └─┬ async@2.0.0-rc.4
│     └── lodash@4.16.1 
└── UNMET PEER DEPENDENCY webpack-dev-server@2.1.0-beta.4
```

**Comment by jbeezley** [comment]
Could be the cache.  I believe Travis keeps seperate caches for each branch.

**Comment by zachmullen** [comment]
It was indeed the cache, I didn't see your comment until this morning, but you and @mgrauer both came up with the solution. Thanks guys!

**Comment by zachmullen** [comment]
@girder/developers any volunteers to QA this?

**Comment by cpatrick** [comment]
I'll take a look.

File: girder/utility/install.py [hunk 1]
```diff
@@ -215,6 +274,9 @@ def main():
 
     web.add_argument('--npm', default="npm",
```
Comment (cpatrick) [inline girder/utility/install.py]: We should fix this quote (not your fault).

File: grunt_tasks/plugin.js [hunk 2]
```diff
@@ -68,105 +78,126 @@ module.exports = function (grunt) {
...
-    grunt.file.expand(grunt.config.get('pluginDir') + '/*')
-        .forEach(function (dir) {
-            var plugin = path.basename(dir);
-            var json = path.resolve(dir, 'plugin.json');
-            var yml = path.resolve(dir, 'plugin.yml');
-            var config = {}, npm;
+        if (fs.existsSync(json)) {
+            config = grunt.file.readYAML(json);
+        }
+        if (fs.existsSync(yml)) {
```
Comment (cpatrick) [inline grunt_tasks/plugin.js]: If both YAML and JSON exist, do we want to override the JSON config with YAML without warning the user?

Comment (cpatrick) [inline grunt_tasks/plugin.js]: I'd say the ordering is fine, but I'd use an `else if` and log a message indicating which is being used. We should also not the order preferences in the documentation.

**Comment by cpatrick** [comment]
I'm getting the following errors when trying to run this on my local dev instance when clicking the magic web_build button:

```
sprintf.js:195 POST http://localhost:9999/api/v1/system/web_build 500 (Internal Server Error)send @ sprintf.js:195ajax @ sprintf.js:195Backbone.ajax @ sprintf.js:195__restRequest @ UploadWidget.js:260restRequest @ UploadWidget.js:260clickGRebuildWebCode @ UploadWidget.js:260dispatch @ sprintf.js:195elemData.handle @ sprintf.js:195
UploadWidget.js:260 500 Internal Server Error {"message": "UnicodeDecodeError: UnicodeDecodeError('ascii', 'girder@1.6.0 /Users/cpatrick/Projects/Kitware/General/girder\\n\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\x80 abbrev@1.0.9 \\n\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\x80 aproba@1.0.4 \\n\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac babel-core@6.13.2\\n\\xe2\\x94\\x82 \\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac babel-register@6.11.6\\n\\xe2\\x94\\x82 \\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\xac home-or-tmp@1.0.0\\n\\xe2\\x94\\x82 \\xe2\\x94\\x82   \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 os-tmpdir@1.0.1 \\n\\xe2\\x94\\x82 \\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac debug@2.2.0 \\n\\xe2\\x94\\x82 \\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 ms@0.7.1 \\n\\xe2\\x94\\x82 \\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac minimatch@3.0.3 \\n\\xe2\\x94\\x82 \\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\xac brace-expansion@1.1.3\\n\\xe2\\x94\\x82 \\xe2\\x94\\x82   \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 concat-map@0.0.1 \\n\\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 path-is-absolute@1.0.0 \\n\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac babel-loader@6.2.4\\n\\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\xac mkdirp@0.5.1 \\n\\xe2\\x94\\x82   \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 minimist@0.0.8 \\n\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac event-source@0.1.1\\n\\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\xac es5-ext@0.10.6\\n\\xe2\\x94\\x82   \\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac es6-iterator@0.1.3\\n\\xe2\\x94\\x82   \\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 d@0.1.1 \\n\\xe2\\x94\\x82   \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\xac es6-symbol@2.0.1\\n\\xe2\\x94\\x82     \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 d@0.1.1 \\n\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac extract-text-webpack-plugin@2.0.0-beta.4\\n\\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 async@1.5.2 \\n\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac fstream@1.0.10 \\n\\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\xac mkdirp@0.5.1 \\n\\xe2\\x94\\x82   \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 minimist@0.0.8 \\n\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\xac fstream-npm@1.2.0 \\n\\xe2\\x94\\x82 \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\xac fstream-ignore@1.0.5 \\n\\xe2\\x94\\x82   \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\xac minimatch@3.0.3 \\n\\xe2\\x94\\x82     \\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\xac brace-expansion@1.1.6 \\n\\xe2\\x94', 61, 62, 'ordinal not in range(128)')", "trace": [["girder/api/rest.py", 505, "endpointDecorator", "val = fun(self, args, kwargs)"], ["girder/api/rest.py", 1054, "POST", "return self.handleRoute(method, path, params)"], ["girder/api/rest.py", 828, "handleRoute", "val = handler(**kwargs)"], ["girder/api/access.py", 39, "wrapped", "return args[0](*iargs, **ikwargs)"], ["girder/api/v1/system.py", 428, "buildWebCode", "install.runWebBuild(dev=dev, progress=progress)"], ["girder/utility/install.py", 148, "runWebBuild", "_pipeOutputToProgress(proc, progress)"], ["girder/utility/install.py", 97, "_pipeOutputToProgress", "buf = buf.decode()"]], "type": "internal"}error @ UploadWidget.js:260fire @ sprintf.js:195fireWith @ sprintf.js:195done @ sprintf.js:195(anonymous function) @ sprintf.js:195
```

**Comment by zachmullen** [comment]
@cpatrick I think I addressed all your comments with those last two commits, PTAL.

**Comment by cpatrick** [comment]
I'm still seeing the same error in the web console.

**Comment by zachmullen** [comment]
Is it the exact same error? I changed the codec, so the error message should hopefully at least slightly have changed...

**Comment by zachmullen** [comment]
Ok, try again, that should actually fix it.

**Comment by cpatrick** [comment]
This did actually fix my issues!

**Comment by zachmullen** [comment]
Thanks for the review!

**Pull request merged**